File: vol3/vol3-csharp-examples/AIFH-Vol3/Examples/Classic/Logic/LogicExample.cs
// Artificial Intelligence for Humans
// Volume 3: Deep Learning and Neural Networks
// C# Version
// http://www.aifh.org
// http://www.jeffheaton.com
//
// Code repository:
// https://github.com/jeffheaton/aifh
//
// Copyright 2015 by Jeff Heaton
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// For more information on Heaton Research copyrights, licenses
// and trademarks visit:
// http://www.heatonresearch.com/copyright
//

using System;
using System.Collections.Generic;

namespace AIFH_Vol3.Examples.Classic.Logic
{
    /// <summary>
    ///     Create a hard-wired (weights directly set) neural network for the logic gates: and, or, not & xor.
    /// </summary>
    public class LogicExample
    {
        /// <summary>
        ///     The name of this example.
        /// </summary>
        public static string ExampleName = "Simple hard-wired neural network for logic gates.";

        /// <summary>
        ///     The chapter this example is from.
        /// </summary>
        public static int ExampleChapter = 1;

        /// <summary>
        ///     Display a truth table and query the neural network.
        /// </summary>
        /// <param name="inputs">The inputs.</param>
        /// <param name="output">The output.</param>
        public static void TruthTable(IList<InputNeuron> inputs, RegularNeuron output)
        {
            var v = new double[inputs.Count];
            var done = false;

            while (!done)
            {
                for (var i = 0; i < inputs.Count; i++)
                {
                    inputs[i].Value = v[i];
                }

                var o = output.Compute();
                Console.WriteLine(v + " : " + o);

                // Roll forward to next row

                var i2 = 0;
                while (i2 < v.Length)
                {
                    v[i2] += 1;
                    if (v[i2] > 1)
                    {
                        v[i2] = 0;
                        i2 += 1;
                    }
                    else
                    {
                        break;
                    }
                }

                if (i2 == v.Length)
                {
                    done = true;
                }
            }
        }

        /// <summary>
        ///     Create a neural network for AND.
        /// </summary>
        public static void ProcessAnd()
        {
            Console.WriteLine("Boolean AND");
            var inputs = new List<InputNeuron>();
            inputs.Add(new InputNeuron());
            inputs.Add(new InputNeuron());


            var output = new RegularNeuron(-1.5);
            output.Parents.Add(new Connection(1, inputs[0]));
            output.Parents.Add(new Connection(1, inputs[1]));
            TruthTable(inputs, output);
        }

        /// <summary>
        ///     Create a neural network for OR.
        /// </summary>
        public static void ProcessOr()
        {
            Console.WriteLine("Boolean OR");
            var inputs = new List<InputNeuron>();
            inputs.Add(new InputNeuron());
            inputs.Add(new InputNeuron());

            var output = new RegularNeuron(-0.5);
            output.Parents.Add(new Connection(1, inputs[0]));
            output.Parents.Add(new Connection(1, inputs[1]));
            TruthTable(inputs, output);
        }

        /// <summary>
        ///     Create a neural network for NOT.
        /// </summary>
        public static void ProcessNot()
        {
            Console.WriteLine("Boolean NOT");
            var inputs = new List<InputNeuron>();
            inputs.Add(new InputNeuron());

            var output = new RegularNeuron(0.5);
            output.Parents.Add(new Connection(-1, inputs[0]));
            TruthTable(inputs, output);
        }

        /// <summary>
        ///     Create a neural network for XOR.
        /// </summary>
        public static void ProcessXor()
        {
            Console.WriteLine("Boolean XOR");
            var inputs = new List<InputNeuron>();
            inputs.Add(new InputNeuron());
            inputs.Add(new InputNeuron());

            var hidden1 = new List<RegularNeuron>();
            hidden1.Add(new RegularNeuron(-0.5));
            hidden1.Add(new RegularNeuron(-1.5));
            hidden1[0].Parents.Add(new Connection(1, inputs[0]));
            hidden1[0].Parents.Add(new Connection(1, inputs[1]));
            hidden1[1].Parents.Add(new Connection(1, inputs[0]));
            hidden1[1].Parents.Add(new Connection(1, inputs[1]));

            var hidden2 = new RegularNeuron(0.5);
            hidden2.Parents.Add(new Connection(-1, hidden1[1]));
            var output = new RegularNeuron(-1.5);
            output.Parents.Add(new Connection(1, hidden1[0]));
            output.Parents.Add(new Connection(1, hidden2));
            TruthTable(inputs, output);
        }


        /// <summary>
        ///     The entry point for this example.  If you would like to make this example
        ///     stand alone, then add to its own project and rename to Main.
        /// </summary>
        /// <param name="args">Not used.</param>
        public static void ExampleMain(string[] args)
        {
            ProcessAnd();
            ProcessOr();
            ProcessNot();
            ProcessXor();
        }
    }
}

File: vol3/vol3-java-examples/src/main/java/com/heatonresearch/aifh/examples/classic/logic/LogicExample.java
/*
 * Artificial Intelligence for Humans
 * Volume 3: Deep Learning and Neural Networks
 * Java Version
 * http://www.aifh.org
 * http://www.jeffheaton.com
 *
 * Code repository:
 * https://github.com/jeffheaton/aifh
 *
 * Copyright 2014-2015 by Jeff Heaton
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * For more information on Heaton Research copyrights, licenses
 * and trademarks visit:
 * http://www.heatonresearch.com/copyright
 */
package com.heatonresearch.aifh.examples.classic.logic;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Create a hard-wired (weights directly set) neural network for the logic gates: and, or, not & xor.
 */
public class LogicExample {

    /**
     * Display a truth table and query the neural network.
     * @param inputs The inputs.
     * @param output The output.
     */
    public static void truthTable(List<InputNeuron> inputs, RegularNeuron output) {
        double[] v = new double[inputs.size()];
        boolean done = false;

        while(!done) {

            for(int i=0;i<inputs.size();i++) {
                inputs.get(i).setValue(v[i]);
            }

            double o = output.compute();
            System.out.println(Arrays.toString(v) + " : " + o);

            // Roll forward to next row

            int i = 0;
            while(i<v.length) {
                v[i] += 1;
                if(v[i] > 1) {
                    v[i] = 0;
                    i += 1;
                } else {
                    break;
                }
            }

            if( i == v.length) {
                done = true;
            }
        }
    }

    /**
     * Create a neural network for AND.
     */
    public static void processAnd() {
        System.out.println("Boolean AND");
        List<InputNeuron> inputs = new ArrayList<>();
        inputs.add(new InputNeuron());
        inputs.add(new InputNeuron());


        RegularNeuron output = new RegularNeuron(-1.5);
        output.getParents().add(new Connection(1,inputs.get(0)));
        output.getParents().add(new Connection(1,inputs.get(1)));
        truthTable(inputs,output);
    }

    /**
     * Create a neural network for OR.
     */
    public static void processOr() {
        System.out.println("Boolean OR");
        List<InputNeuron> inputs = new ArrayList<>();
        inputs.add(new InputNeuron());
        inputs.add(new InputNeuron());


        RegularNeuron output = new RegularNeuron(-0.5);
        output.getParents().add(new Connection(1,inputs.get(0)));
        output.getParents().add(new Connection(1,inputs.get(1)));
        truthTable(inputs,output);
    }

    /**
     * Create a neural network for NOT.
     */
    public static void processNot() {
        System.out.println("Boolean NOT");
        List<InputNeuron> inputs = new ArrayList<>();
        inputs.add(new InputNeuron());

        RegularNeuron output = new RegularNeuron(0.5);
        output.getParents().add(new Connection(-1,inputs.get(0)));
        truthTable(inputs,output);
    }

    /**
     * Create a neural network for XOR.
     */
    public static void processXor() {
        System.out.println("Boolean XOR");
        List<InputNeuron> inputs = new ArrayList<>();
        inputs.add(new InputNeuron());
        inputs.add(new InputNeuron());

        List<RegularNeuron> hidden1 = new ArrayList<>();
        hidden1.add(new RegularNeuron(-0.5));
        hidden1.add(new RegularNeuron(-1.5));
        hidden1.get(0).getParents().add(new Connection(1,inputs.get(0)));
        hidden1.get(0).getParents().add(new Connection(1,inputs.get(1)));
        hidden1.get(1).getParents().add(new Connection(1,inputs.get(0)));
        hidden1.get(1).getParents().add(new Connection(1,inputs.get(1)));

        RegularNeuron hidden2 = new RegularNeuron(0.5);
        hidden2.getParents().add(new Connection(-1,hidden1.get(1)));
        RegularNeuron output = new RegularNeuron(-1.5);
        output.getParents().add(new Connection(1,hidden1.get(0)));
        output.getParents().add(new Connection(1,hidden2));
        truthTable(inputs,output);
    }


    /**
     * Program entry point.
     * @param args Not used.
     */
    public static void main(String[] args) {
        processAnd();
        processOr();
        processNot();
        processXor();

    }
}
